src/Terrain/MapWindow.java

```diff
@@ -53,13 +64,22 @@ public Environment getEnvironment(int w, int h)
   /** Returns the Location at the specified location, or null if that Location does not exist. */
   public Location getLocation(XYCoord location)
   {
-    return master.getLocation(location);
+    return getLocation(location.xCoord, location.yCoord);
   }
 
   /** Returns the Location at the specified location, or null if that Location does not exist. */
-  public Location getLocation(int w, int h)
+  public Location getLocation(int x, int y)
   {
-    return master.getLocation(w, h);
+    XYCoord coord = new XYCoord(x, y);
+    Location masterLoc = master.getLocation(coord);
+    Location returnLoc = masterLoc;
+    if( isLocationFogged(coord) )
+    {
+      returnLoc = new Location(returnLoc.getEnvironment(), coord);
```

Comment:
The cloning solution did occur to me, but I didn't feel like outlawing Locations from being used as the keys in maps.
I guess XYCoord works just as well in most contexts, though.
Especially now that every CO and thus every unit now technically can get you full game information, heh.